src/Data/Bifunctor/ShowRead.hs

```diff
@@ -0,0 +1,116 @@
+{-# language ConstraintKinds #-}
+{-# language DataKinds #-}
+{-# language FlexibleInstances #-}
+{-# language FunctionalDependencies #-}
+{-# language InstanceSigs #-}
+{-# language KindSignatures #-}
+{-# language MultiParamTypeClasses #-}
+{-# language QuantifiedConstraints #-}
+{-# language ScopedTypeVariables #-}
+{-# language TypeApplications #-}
+{-# language TypeFamilies #-}
+{-# language UndecidableInstances #-}
+{-# language Safe #-}
+
+-- | Types for lifting instances of `Show`N and `Read`N
+-- for record newtypes. We don't show record syntax, because
+-- it's too much clutter, but we accept it when reading.
+module Data.Bifunctor.ShowRead
+  ( ShowRead (..)
+  , ShowRead1 (..)
+  , ShowRead2 (..)
+  , liftReadPrecWhatever
+  , liftShowsPrecWhatever
+  ) where
+import qualified Text.ParserCombinators.ReadPrec as TPR
+import qualified Text.Read.Lex as TRL
+import qualified Text.Read as TR
+import Text.Read (ReadPrec, Read (..), readListPrecDefault)
+import GHC.Generics
+import Data.Kind
+import Data.Functor.Classes
+
+newtype ShowRead a = ShowRead a
+newtype ShowRead1 f a = ShowRead1 (f a)
+newtype ShowRead2 f a b = ShowRead2 (f a b)
+
+instance (Wraps n d c s o, Read o) => Read (ShowRead n) where
+  readPrec = ShowRead <$> liftReadPrecWhatever readPrec
+  readListPrec = readListPrecDefault
+
+instance (Wraps n d c s o, Show o) => Show (ShowRead n) where
+  showsPrec d (ShowRead x) = liftShowsPrecWhatever showsPrec d x
+
+instance (Wraps1 n d c s o, Read1 o, Wraps (n ()) d c s (o ())) => Read1 (ShowRead1 n) where
+  liftReadPrec :: forall a. ReadPrec a -> ReadPrec [a] -> ReadPrec (ShowRead1 n a)
+  liftReadPrec rp rl = ShowRead1 <$> liftReadPrecWhatever @(n a) @d @c @s @(o a) (liftReadPrec rp rl)
```

Comment:
Why do I need to specify every single type here?

Reply:
I'm not seeing ambiguity *there*.

```haskell
type WrapsF n d c s o =
  ( Generic n
  , Rep n ~ D1 d (C1 c (S1 s (Rec0 o)))
  , Constructor c
  , Selector s )

class WrapsF n d c s o => Wraps n d c s o
instance WrapsF n d c s o => Wraps n d c s o

liftReadPrecWhatever
  :: forall n d c s o.
     ( Generic n
     , Wraps n d c s o
     , Constructor c
     , Selector s)
  => ReadPrec o -> ReadPrec n
```

The `n` in the result type should determine all the other types through `Rep n ~ D1 d (C1 c (S1 s (Rec0 o)))`. I'm guessing there's something to do with the quantified constraints going on. Speaking of which, I don't really understand why I seem to need those `Wraps (n ()) d c s (o ())` and so on, which are truly ugly. I can't help thinking that either I'm doing something more deeply incorrectly or something in GHC is busted.

Reply:
I've never been able to fully internalize the rules GHC uses to determine when something is considered ambiguous or not, and it wouldn't surprise me if there were some dark corners involving quantified constraints. I don't have a good alternative for the `Wraps (n ()) d c s (o ())` trick, either.